File: Easy/Remove Duplicate Letters.py
class Solution:
    def removeDuplicateLetters(self, s: str) -> str:
        result = []
        frequency = Counter(s)
        added = set()
        for ch in s:
            frequency[ch] -=1
            if ch in added:
                continue
            while result and result[-1] > ch and frequency[result[-1]]>0:
                added.remove(result.pop())
            if ch not in added:
                result.append(ch)
                added.add(ch)
        return (''.join(result))

# Runtime: 32 ms, faster than 89.59% of Python3 online submissions for Remove Duplicate Letters.
# Memory Usage: 14.2 MB, less than 100.00% of Python3 online submissions for Remove Duplicate Letters.

File: Easy/RemoveDuplicateLetters.java
import java.util.HashMap;
import java.util.HashSet;
import java.util.Set;
import java.util.Stack;

class Solution {
    public String smallestSubsequence(String s) {
        Set<Character> added = new HashSet<Character>();
        HashMap<Character, Integer> frequency= new HashMap<Character, Integer>();
        Stack<Character> result = new Stack<Character>();

        for (Character ch: s.toCharArray()){
            Integer value = frequency.get(ch);
            if (value == null){
                frequency.put(ch, 1);
            } else {
                frequency.put(ch, value+1);
            }
        }

        for (Character ch: s.toCharArray()){
            frequency.put(ch, frequency.get(ch)-1);
            if (added.contains(ch)){
                continue;
            }
            while (!result.isEmpty() && frequency.get(result.lastElement())> 0 && result.lastElement()> ch) {
                added.remove(result.pop());
            }
            added.add(ch);
            result.push(ch);
        }
        String finalResult = new String();
        while (!result.isEmpty()){
            finalResult = result.pop()+finalResult;
        }
        return finalResult;
    }
}